File: day02/Puzzle.cs
using Range = (long First, long Last);

internal class Puzzle(string rawInput) : AocPuzzle<Range[], long>(rawInput)
{
    private static Range CreateRange(string line)
    {
        var parts = line.Split('-');
        return new(long.Parse(parts[0]), long.Parse(parts[1]));
    }

    protected override Range[] ParseInput(string rawInput) => [..rawInput.Split(',').Select(CreateRange)];

    protected override long RunPartOne()
    {
        long result = 0;
        foreach (var (first, last) in _input)
        {
            (string firstString, string lastString) = (first.ToString(), last.ToString());
            if ((firstString.Length & 1) + (lastString.Length & 1) > 1)
                continue;

            for (long i = first; i <= last; ++i)
            {
                var curSpan = i.ToString().AsSpan();
                if ((curSpan.Length & 1) > 0)
                    continue;

                int midpoint = curSpan.Length / 2;
                if (curSpan[..midpoint].Equals(curSpan[midpoint..], StringComparison.Ordinal))
                    result += i;
            }
        }
        return result;
    }

    protected override long RunPartTwo()
    {
        long result = 0;
        foreach (var (first, last) in _input)
        {
            for (long i = first; i <= last; ++i)
            {
                var curSpan = i.ToString().AsSpan();
                int midpoint = curSpan.Length / 2;
                for (int j = 1; j <= midpoint; ++j)
                {
                    if (curSpan.Length % j > 0)
                        continue;

                    var pattern = curSpan[..j];
                    bool match = true;
                    for (int k = j; k < curSpan.Length; k += j)
                    {
                        if (!curSpan[k..(k + j)].Equals(pattern, StringComparison.Ordinal))
                        {
                            match = false;
                            break;
                        }
                    }

                    if (match)
                    {
                        result += i;
                        break;
                    }
                }
            }
        }
        return result;
    }
}

File: day05/Puzzle.fs
module Puzzle

type Range = int64 * int64
type Input = { Ranges: Range array; Ingredients: int64 array }

let private createRange (line: string) =
    let parts = line.Split('-')
    int64 parts[0], int64 parts[1]

let private isInRange value range = value >= fst range && value <= snd range

let private count condition items =
    items
    |> Array.sumBy (fun item -> if condition item then 1L else 0)

let private countUnique ranges =
    let rec loop count maxValue curRanges =
        match curRanges with
        | [] -> count
        | range :: nextRanges ->
            let curMax = snd range
            if curMax <= maxValue then
                loop count maxValue nextRanges
            else
                let curCount =
                    if isInRange maxValue range then
                        curMax - maxValue
                    else
                        curMax - fst range + 1L

                loop (count + curCount) curMax nextRanges

    loop 0 0L (ranges |> Array.sort |> Array.toList)

let parseInput (rawInput: string) =
    let sections = rawInput.Split("\n\n") |> Array.map _.Split('\n')
    { Ranges = sections[0] |> Array.map createRange
      Ingredients = sections[1] |> Array.map int64 }

let runPartOne (input: Input) =
    input.Ingredients
    |> count (fun ingredient -> input.Ranges |> Array.exists (isInRange ingredient))

let runPartTwo (input: Input) = input.Ranges |> countUnique
